src/util/simplify_expr.cpp

```diff
@@ -784,16 +790,45 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
         expr=fixedbv.to_expr();
         return false;
       }
+      else if(expr_type_id == ID_bv)
+      {
+        expr = from_integer(f.pack(), expr_type);
+        return false;
+      }
     }
     else if(op_type_id==ID_bv)
     {
-      if(expr_type_id==ID_unsignedbv ||
-         expr_type_id==ID_signedbv ||
-         expr_type_id==ID_floatbv)
+      if(
+        expr_type_id == ID_unsignedbv || expr_type_id == ID_signedbv ||
+        expr_type_id == ID_c_enum || expr_type_id == ID_c_enum_tag ||
+        expr_type_id == ID_c_bit_field)
       {
         const auto width = to_bv_type(op_type).get_width();
         const auto int_value = bvrep2integer(value, width, false);
-        expr=from_integer(int_value, expr_type);
+        if(expr_type_id != ID_c_enum_tag)
+          expr = from_integer(int_value, expr_type);
+        else
+          expr = from_integer(
+            int_value, ns.follow_tag(to_c_enum_tag_type(expr_type)));
```

Comment:
I am afraid that the tag type needs to be restored, or otherwise the type of the expression changes.

Reply:
Err, thanks for catching that!